### PR TITLE
drivers/soft_spi: fix soft spi transfer bug

### PR DESCRIPTION
**Pull request by kex2017**
fix:fix drivers/soft_spi transfer one byte without read back bug
test:test read w25q64 id success

### CODE REVIEW

**Comment by kfessel** [comment]
I am no reviewer but to interested by the Tietel of your PR

The following might be early unnecessary optimization and i currently have no spi device to check but i would go with this:

out as a transfer register; shifting only by one (some archs (e.g.:AVR) a bad at shifting more than on step at once)

```
static inline uint8_t _transfer_one_byte(soft_spi_t bus, uint8_t out)
{
    int8_t bit = 0, i = 8;
    if (SOFT_SPI_MODE_1 == soft_spi_config[bus].soft_spi_mode ||
        SOFT_SPI_MODE_3 == soft_spi_config[bus].soft_spi_mode) {
        /* CPHA = 1*/
        gpio_toggle(soft_spi_config[bus].clk_pin);
    }

    do{
        bit = (out & (1 << 7)) >> 7;
        gpio_write(soft_spi_config[bus].mosi_pin, bit);

        xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
        gpio_toggle(soft_spi_config[bus].clk_pin);

        out <<= 1; /*shift tranfer register*/

        bit = gpio_read(soft_spi_config[bus].miso_pin);
        out = bit ? out | 0x01 : out & 0xfe; /*set or delete bit 0*/

        xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
        --i;
        if (i > 0) gpio_toggle(soft_spi_config[bus].clk_pin);
    }while(i > 0)

    if (SOFT_SPI_MODE_0 == soft_spi_config[bus].soft_spi_mode ||
        SOFT_SPI_MODE_2 == soft_spi_config[bus].soft_spi_mode) {
        /* CPHA = 0 */
        xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
        gpio_toggle(soft_spi_config[bus].clk_pin);
    }

    return out;
}
```

For the read function i would just transfer 0x00 as done for hardware SPI.

**Comment by kex2017** [comment]
> I am no reviewer but to interested by the Tietel of your PR
> 
> The following might be early unnecessary optimization and i currently have no spi device to check but i would go with this:
> 
> out as a transfer register; shifting only by one (some archs (e.g.:AVR) a bad at shifting more than on step at once)
> 
> ```
> static inline uint8_t _transfer_one_byte(soft_spi_t bus, uint8_t out)
> {
>     int8_t bit = 0, i = 8;
>     if (SOFT_SPI_MODE_1 == soft_spi_config[bus].soft_spi_mode ||
>         SOFT_SPI_MODE_3 == soft_spi_config[bus].soft_spi_mode) {
>         /* CPHA = 1*/
>         gpio_toggle(soft_spi_config[bus].clk_pin);
>     }
> 
>     do{
>         bit = (out & (1 << 7)) >> 7;
>         gpio_write(soft_spi_config[bus].mosi_pin, bit);
> 
>         xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
>         gpio_toggle(soft_spi_config[bus].clk_pin);
> 
>         out <<= 1; /*shift tranfer register*/
> 
>         bit = gpio_read(soft_spi_config[bus].miso_pin);
>         out = bit ? out | 0x01 : out & 0xfe; /*set or delete bit 0*/
> 
>         xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
>         --i;
>         if (i > 0) gpio_toggle(soft_spi_config[bus].clk_pin);
>     }while(i > 0)
> 
>     if (SOFT_SPI_MODE_0 == soft_spi_config[bus].soft_spi_mode ||
>         SOFT_SPI_MODE_2 == soft_spi_config[bus].soft_spi_mode) {
>         /* CPHA = 0 */
>         xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
>         gpio_toggle(soft_spi_config[bus].clk_pin);
>     }
> 
>     return out;
> }
> ```
> 
> For the read function i would just transfer 0x00 as done for hardware SPI.

test on w25q64 ok

**Comment by kfessel** [comment]
@kex2017: Thanks for testing

Even though I am not a reviewer, i would suggest:

- do not PR from your master branch, create a new one prior PR `git checkout -b <my new branch>` and `git push -u <your_repo> <my_new_branch>`
- keep the Headlines that you got when you started your PR (try to fit your information within them)
- do not merge master into your PR instead someone may ask you to `git rebase` on <RIOT/origin>/master
- the static test already have some issues with spelling and code formating (sorry for that; these tests don't like my writing style)
you might want to 'git commit --fixup'
- If you want to restart the PR please put this one into references

**Comment by kex2017** [comment]
Thank you for your advice

**Comment by kfessel** [comment]
@basilfx and @maribu might have interest at this PR that basicaly adds read support to soft_spi

File: drivers/soft_spi/soft_spi.c [hunk 1]
```diff
@@ -137,29 +138,33 @@ void soft_spi_release(soft_spi_t bus)
 
 static inline uint8_t _transfer_one_byte(soft_spi_t bus, uint8_t out)
 {
-    int8_t bit = 0, i = 0;
+    int8_t i = 8;
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:141]:
```suggestion
    uint8_t i = 8;
```

It is better to use `unsigned` types for bit operations, as C allows for [arithmetic shifts](https://en.wikipedia.org/wiki/Arithmetic_shift) to be used on signed types.

File: drivers/soft_spi/soft_spi.c [hunk 2]
```diff
@@ -137,29 +138,33 @@ void soft_spi_release(soft_spi_t bus)
...
         SOFT_SPI_MODE_3 == soft_spi_config[bus].soft_spi_mode) {
         /* CPHA = 1*/
         gpio_toggle(soft_spi_config[bus].clk_pin);
     }
 
-    bit = (out & (1 << 7)) >> 7;
-    gpio_write(soft_spi_config[bus].mosi_pin, bit);
-    for (i = 6; i >= 0; i--) {
+    do{
+        int8_t bit = (out & (1 << 7)) >> 7;
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:149]:
```suggestion
        uint8_t bit = out >> 7;
```

File: drivers/soft_spi/soft_spi.c [hunk 3]
```diff
@@ -137,29 +138,33 @@ void soft_spi_release(soft_spi_t bus)
...
+        gpio_write(soft_spi_config[bus].mosi_pin, bit);
+
         xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
         gpio_toggle(soft_spi_config[bus].clk_pin);
+
+        out <<= 1; /*shift transfer register*/
+
+        bit = gpio_read(soft_spi_config[bus].miso_pin);
+        out = bit ? out | 0x01 : out & 0xfe; /*set or delete bit 0*/
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:158]:
```suggestion
        out = bit ? (out | 0x01) : (out & 0xfe); /*set or delete bit 0*/
```

This is personal taste, so feel free to ignore. But IMO the extra braces here ease reading.

File: drivers/soft_spi/soft_spi.c [hunk 4]
```diff
@@ -137,29 +138,33 @@ void soft_spi_release(soft_spi_t bus)
...
+
         xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
-        gpio_toggle(soft_spi_config[bus].clk_pin);
-        bit = (out & (1 << i)) >> i;
-        gpio_write(soft_spi_config[bus].mosi_pin, bit);
-    }
-    xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
-    gpio_toggle(soft_spi_config[bus].clk_pin);
+        --i;
+        if (i > 0) gpio_toggle(soft_spi_config[bus].clk_pin);
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:162]:
```suggestion
        if (i > 0) {
            gpio_toggle(soft_spi_config[bus].clk_pin);
        }
```

File: drivers/soft_spi/soft_spi.c [hunk 5]
```diff
@@ -137,29 +138,33 @@ void soft_spi_release(soft_spi_t bus)
...
+    }while(i > 0);
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:163]:
```suggestion
    } while (i > 0);
```

File: drivers/soft_spi/soft_spi.c [hunk 6]
```diff
@@ -191,22 +196,49 @@ uint8_t soft_spi_transfer_byte(soft_spi_t bus, soft_spi_cs_t cs, bool cont, uint
...
 
-    for (size_t i = 0; i < len-1; i++) {
-        tmp = (NULL != out) ? ((uint8_t *)out)[i] : 0;
-        uint8_t retval = soft_spi_transfer_byte(bus, cs, true, tmp);
-        if (NULL != in) {
-            ((uint8_t *)in)[0] = retval;
+    if (!inbuf)
+    {
+        for (size_t i = 0; i < len; i++)
+        {
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:218]:
```suggestion
    if (!inbuf) {
        for (size_t i = 0; i < len; i++) {
```

File: drivers/soft_spi/soft_spi.c [hunk 7]
```diff
@@ -191,22 +196,49 @@ uint8_t soft_spi_transfer_byte(soft_spi_t bus, soft_spi_cs_t cs, bool cont, uint
...
+            _transfer_one_byte(bus, outbuf[i]);
+        }
+    }
+    else if (!outbuf)
+    {
+        for (size_t i = 0; i < len; i++)
+        {
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:225]:
```suggestion
    else if (!outbuf) {
        for (size_t i = 0; i < len; i++) {
```

File: drivers/soft_spi/soft_spi.c [hunk 8]
```diff
@@ -191,22 +196,49 @@ uint8_t soft_spi_transfer_byte(soft_spi_t bus, soft_spi_cs_t cs, bool cont, uint
...
+            inbuf[i] = _transfer_one_byte(bus, READ_PADDING_BYTE);
+        }
+    }
+    else
+    {
+        for (size_t i = 0; i < len; i++)
+        {
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:232]:
```suggestion
    else  {
        for (size_t i = 0; i < len; i++)  {
```

File: drivers/soft_spi/soft_spi.c [hunk 9]
```diff
@@ -191,22 +196,49 @@ uint8_t soft_spi_transfer_byte(soft_spi_t bus, soft_spi_cs_t cs, bool cont, uint
...
+            inbuf[i] = _transfer_one_byte(bus, outbuf[i]);
         }
     }
 
-    tmp = (NULL != out) ? ((uint8_t *)out)[len-1] : 0;
-    soft_spi_transfer_byte(bus, cs, cont, tmp);
+    if (!cont)
+    {
+        if ((cs != GPIO_UNDEF) && (cs != SOFT_SPI_CS_UNDEF))
+        {
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:240]:
```suggestion
    if (!cont)  {
        if ((cs != GPIO_UNDEF) && (cs != SOFT_SPI_CS_UNDEF))  {
```

File: drivers/soft_spi/soft_spi.c [hunk 10]
```diff
@@ -191,22 +196,49 @@ uint8_t soft_spi_transfer_byte(soft_spi_t bus, soft_spi_cs_t cs, bool cont, uint
...
+    /* make sure at least one input or one output buffer is given */
+    assert(out || in);
+
+    const uint8_t *outbuf = out;
+    uint8_t *inbuf = in;
 
-    uint8_t tmp = 0;
+    /* activate the given chip select line */
+    if ((cs != GPIO_UNDEF) && (cs != SOFT_SPI_CS_UNDEF))
+    {
```
Comment (maribu) [inline drivers/soft_spi/soft_spi.c:211]:
```suggestion
    if ((cs != GPIO_UNDEF) && (cs != SOFT_SPI_CS_UNDEF)) {
```

**Review by maribu — COMMENTED** [review]
Thanks for this fix. I wonder how this ever got merged in this state.

I have some comments regarding code style.

File: drivers/soft_spi/soft_spi.c (hunk 1 above)
Comment (kex2017) [reply drivers/soft_spi/soft_spi.c]: I'll take your advice

**Comment by maribu** [comment]
Code looks good to me. Let's see if the CI has some comments. Feel free to squash.

It might be that now some boards fail to build test applications or similar due to the increased ROM size. If that happens, we need to disable the final linking stage in the affected application by adding the affected board to the applications `Makefile.ci` - but maybe we're like and everything still builds fine.

**Comment by kex2017** [comment]
> adding the affected board to the applications `Makefile.ci`

Maybe add the affected board to the applications Makefile.ci is a good way

**Comment by maribu** [comment]
> > adding the affected board to the applications `Makefile.ci`
> 
> Maybe add the affected board to the applications Makefile.ci is a good way

Not needed, all builds succeeded :-)

**Comment by maribu** [comment]
Tested using https://github.com/RIOT-OS/RIOT/pull/16196. Please squash!

**Comment by maribu** [comment]
@kex2017 Could you combine all commits into the first using `git rebase -i master`? Then change `pick` to `fixup` for all but the first commit. Here is [a random search hit for a guide on squashing](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) for more details.

**Comment by kfessel** [comment]
@kex2017 maybe its not so easy for you since this is your master branch you need to rebase it on upstream/master 

`git rebase -i origin/master`  if origin is the upstream RIOT remote

if you do not yet have an upstream RIOT remote you might need to create one 

`git remote add up https://github.com/kex2017/RIOT.git`
`git fetch up`

and rebase on its master 

`git rebase -i up/master` 

and then mark your commits as @maribu  described

**Review by maribu — APPROVED** [review]
ACK. Thanks for the fix. A very nice first contribution, btw!

**Pull request merged**